File: go/300.600/378.KthSmallestElementInASortedMatrix.go
package main

import (
	"fmt"
	"math"
)

func kthSmallest(matrix [][]int, k int) int {
	dp := make([]int, len(matrix))
	ret := 0
	for ; k > 0; k-- {
		curr := 0
		ret = math.MaxInt32
		for i, v := range dp {
			if v < len(matrix[0]) && matrix[i][v] < ret {
				ret = matrix[i][v]
				curr = i
			}
		}
		dp[curr]++
	}
	return ret
}

func kthSmallest1(matrix [][]int, k int) int {
	left := matrix[0][0]
	right := matrix[len(matrix)-1][len(matrix[0])-1]
	for left < right {
		mid := left + (right-left)/2
		if count(matrix, mid) >= k {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return left
}

func count(matrix [][]int, mid int) int {
	i := len(matrix) - 1
	j := 0
	num := 0
	for i >= 0 && j < len(matrix[0]) {
		if matrix[i][j] <= mid {
			num += i + 1
			j++
		} else {
			i--
		}
	}
	return num
}

func main() {
	fmt.Println(kthSmallest1([][]int{{-5, -4}, {-5, -4}}, 2))
	fmt.Println(kthSmallest([][]int{{1, 5, 9}, {10, 11, 13}, {12, 13, 15}}, 8), kthSmallest1([][]int{{1, 5, 9}, {10, 11, 13}, {12, 13, 15}}, 8), 13)
	fmt.Println(kthSmallest([][]int{{-5}}, 1), kthSmallest1([][]int{{-5}}, 1), -5)
	fmt.Println(kthSmallest([][]int{{1, 3, 5}, {6, 7, 12}, {11, 14, 14}}, 3), kthSmallest1([][]int{{1, 3, 5}, {6, 7, 12}, {11, 14, 14}}, 3), 5)
}

File: 300.600/378.KthSmallestElementInASortedMatrix.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <functional>
using namespace std;

class KthSmallestElementInASortedMatrix{
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        //优先队
        priority_queue<int> que;
        for(auto& row : matrix){
            for(auto c : row){
                que.push(c);
            }
        }
        while(que.size() > k){
            que.pop();
        }
        return que.top();
    }
public:
    int kthSmallest1(vector<vector<int>>& matrix, int k) {
        //按行排序
        vector<int> dp(matrix.size(), 0);
        int size = matrix[0].size();
        int ret = 0;
        while (k > 0)
        {
            int curr = 0;
            ret = INT_MAX;
            for(int i = 0; i< matrix.size();++i){
                if(dp[i] < size && matrix[i][dp[i]] < ret){
                    curr = i;
                    ret = matrix[i][dp[i]];
                }
            }
            dp[curr]+=1;
            --k;
        }
        return ret;
    }
public:
    int kthSmallest2(vector<vector<int>>& matrix, int k) {
        //二分统计法
        int left = matrix[0][0];
        int right = matrix.back().back();
        while(left < right){
            int mid = left + (right - left) / 2;
            if(count(matrix, mid) >= k){
                right = mid;
            }else{
                left = mid+1;
            }
        }
        return left;
    }
private:
    int count(vector<vector<int>>& matrix, int mid){
        int i = matrix.size()-1;
        int j = 0;
        int num = 0;
        while (i >= 0 && j < matrix[0].size()){
            if(matrix[i][j] <= mid){
                num += i+1;
                ++j;
            }else{
                --i;
            }
        }
        return num;
    }
};

int main(int argc, char const *argv[]){
    KthSmallestElementInASortedMatrix cls;
    cout << cls.kthSmallest2(vector<vector<int>>{{-5, -4}, {-5, -4}}, 2) << " -> -5" << endl;
    cout << cls.kthSmallest2(vector<vector<int>>{{1,5,9},{10,11,13},{12,13,15}}, 8) << " -> 13" << endl;
    cout << cls.kthSmallest2(vector<vector<int>>{{-5}}, 1) << " -> -5" << endl;
    cout << cls.kthSmallest2(vector<vector<int>>{{1,3,5},{6,7,12},{11,14,14}}, 3) << " -> 5" << endl;
    return 0;
}
